### PR TITLE
[Experimental] Enable multiobjective within optimizers

### PR DESCRIPTION
**Pull request by jrapin**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by teytaud** [comment]
...good luck, that one is super scary...

File: nevergrad/optimization/multiobjective/test_core.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        values.append(hvol.add(param))
+    expected = [10, 10, -400, -2500.0, -2500.0, -2470.0, -3300.0, -4100.0]
+    assert values == expected, f"Expected {expected} but got {values}"
+    front = [p.value for p in hvol.pareto_front()]
+    expected_front = [(50, 50), (30, 60), (60, 30)]
+    assert front == expected_front, f"Expected {expected_front} but got {front}"
+
+
+# pylint: disable=redefined-outer-name,unsubscriptable-object,unused-variable,unused-import,reimported,import-outside-toplevel
+def test_doc_multiobjective() -> None:
```
Comment (jrapin) [inline nevergrad/optimization/multiobjective/test_core.py:34]: @teytaud below is an example of how that could work

File: nevergrad/optimization/multiobjective/core.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            # This point is not on the front
+            # First we prune.
+            self._filter_pareto_front()
+            distance_to_pareto = float("Inf")
+            for param in self._pareto:
+                stored_losses = param.loss
+                # TODO the following is probably not good at all:
+                # -> +inf if no point is strictly better (but lower if it is)
+                if (stored_losses <= losses).all():
+                    distance_to_pareto = min(distance_to_pareto, min(losses - stored_losses))
```
Comment (jrapin) [inline nevergrad/optimization/multiobjective/core.py:85]: @teytaud any thought on this?
```python
# TODO the following is probably not good at all:
# -> +inf if no point is strictly better (but lower if it is)
```

File: nevergrad/optimization/multiobjective/core.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        """
+        if not isinstance(parameter, ng.p.Parameter):
+            raise TypeError(f"{self.__class__.__name__}.add should receive a ng.p.Parameter, but got: {parameter}.")
+        losses = parameter.loss
+        if not isinstance(losses, np.ndarray):
+            raise TypeError(f"Parameter should have multivalue as loss, but parameter.loss={losses} ({type(losses)}).")
+        if self._auto_bound > 0:
+            self._auto_bound -= 1
+            if (self._upper_bounds > -float("inf")).all() and (losses > self._upper_bounds).all():
+                return float('inf')  # Avoid uniformly worst points
```
Comment (jrapin) [inline nevergrad/optimization/multiobjective/core.py:59]: @teytaud I've added this case to auto-bound:
if one point is uniformly bad compared to previous ones, do not include it
that makes the example work way better. Is it a problem though?

File: nevergrad/optimization/multiobjective/core.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    -----
+    - This function is not stationary!
+    - The minimum value obtained for this objective function is -h,
+      where h is the hypervolume of the Pareto front obtained, given upper_bounds as a reference point.
+    - The callable keeps track of the pareto_front (see attribute paretor_front) and is therefor stateful.
+      For this reason it cannot be distributed. A user can however call the multiobjective_function
+      remotely, and aggregate locally. This is what happens in the "minimize" method of optimizers.
+    """
+
+    def __init__(self, upper_bounds: tp.Optional[tp.ArrayLike] = None, auto_bound: int = 15) -> None:
```
Comment (jrapin) [inline nevergrad/optimization/multiobjective/core.py:34]: the auto-bound number should probably depend on the dimension :s

**Comment by jrapin** [comment]
Todo:
[ ] integration with Experiments
[ ] deal with optimizers working with param.loss
[ ] remove multi obj descriptor in NGO?

**Pull request merged**